include/seqan3/alignment/cigar_conversion/cigar_from_alignment.hpp

```diff
@@ -0,0 +1,236 @@
+ *
+ * The extended cigar string would look like this:
+ *
+ * `[('=',3)('X',1)('I',2)('=',3)('X',1)('=',1)('D',2)('=',1)]`.
+ *
+ * \include test/snippet/alignment/cigar_conversion/cigar_from_alignment.cpp
+ *
+ * ### Soft and Hard clipping
+ *
+ * The terms soft and hard clipping were introduced by the
+ * [SAM specifications](https://samtools.github.io/hts-specs/SAMv1.pdf). A SAM file is only storing a semi alignment
+ * represented by the CIGAR string. the semi alignment of a query sequence is most often the result of a
+ * [read mapping.](https://training.galaxyproject.org/training-material/topics/sequence-analysis/tutorials/mapping/tutorial.html)
+ *
+ * #### Hard clipping
+ *
+ * Before aligning a query or read to the reference, some tools crop the query sequence because the quality is bad at
+ * one end (e.g. illumina read sometime feature a bad sequence quality towards the end of the read).
+ *
+ * To inform the user of a SAM file that query sequences were altered, hard-clipping information is appended to the
+ * CIGAR string. E.g. `100M50H` indicates that a read of former length `150` has been cropped at the right end by `50`
+ * bases. The sequence in the SAM file will thus only be 100 bases long.
+ *
+ * #### Soft clipping
+ *
+ * In constrast to hard clipping, a soft clipping indicated that the read was that altered/cropped but the respective
+ * bases still do not particiapte in the alignment.
+ * E.g. `100M50S` indicates that a read of length `150` has been aligned without the rightmost `50` bases
+ * The sequence in the SAM file will thus be 150 bases long.
+ *
+ * #### Adding soft and hard clipping with `seqan::cigar_from_alignment`
+ *
+ * You can add the respective clipping information by passing an instance of seqan3::cigar_clipped_bases:
+ *
+ * \include test/snippet/alignment/cigar_conversion/cigar_from_alignment_with_clipping.cpp
+ *
+ * \sa seqan3::cigar_clipped_bases
+ * \sa seqan3::sam_file_output
+ * \sa seqan3::cigar
+ * \sa seqan3::aligned_sequence
+ */
+template <typename alignment_type>
+inline auto cigar_from_alignment(alignment_type const & alignment,
+                                 cigar_clipped_bases const & clipped_bases = {},
+                                 bool const extended_cigar = false)
+{
+    static_assert((tuple_like<std::remove_cvref_t<alignment_type>>
+                   && std::tuple_size_v<std::remove_cvref_t<alignment_type>> == 2),
+                  "The alignment must be a std::pair or std::tuple of size 2.");
+
+    static_assert(
+        (std::equality_comparable_with<gap, std::ranges::range_reference_t<decltype(std::get<0>(alignment))>>
+         && std::equality_comparable_with<gap, std::ranges::range_reference_t<decltype(std::get<1>(alignment))>>),
+        "The alignment must contain two ranges whose value_type is comparable to seqan3::gap");
+
+    /*lambda brief: Compares two seqan3::aligned_sequence values and returns their cigar operation.
```

Comment:
What is `lambda brief`?

Reply:
I copied this over from a free function in IO and made it into a lambda. Usually we do not document lambdas. But it felt bad to remove all of this valuable documentation. so I called this the lambda brief, as in the brief of the lambda.. Looks like it is confusing :D

I guess this
```suggestion
    /*brief: Compares two seqan3::aligned_sequence values and returns their cigar operation.
```
Will be better and without the `!` doxygen will ignore it.

Reply:
Ah so its not expected to appear in the doxygen docu. Yes maybe your suggestion is less confusing :)